src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateActions = [];
+  const stateCopy = {
+    ...state,
+  };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clearProperties(stateCopy);
+        break;
+
+      default:
+        return `Error, unknown action type: ${action.type}`;
+    }
+    stateActions.push({ ...stateCopy });
+  }
+
+  return stateActions;
+}
+
+function addProperties(stateCopy, extraData) {
+  Object.assign(stateCopy, extraData);
+}
```

Comment:
[NAMING] - The function `addProperties` mutates the `stateCopy` directly, which is fine, but the name `stateCopy` suggests that it's a copy that shouldn't be mutated. Consider renaming it to something that indicates it's the current state being worked on, like `currentState`.